### PR TITLE
feat(agent-integrator): convert tools list to object for opencode target

### PR DESCRIPTION
**Pull request by futbolsalas15**
## TL;DR

OpenCode expects the `tools` frontmatter field in agent/skill files to be an **object** (`{Read: true, Glob: true}`).  Some agent files — especially those authored for other harnesses — define it as a **list** (e.g. `[Read, Glob]`) or a comma-separated string.  When `apm install` copies such files to `.opencode/agents/` verbatim, opencode rejects them with a schema validation error.

## What this PR does

1. Adds `_write_opencode_agent()` to `AgentIntegrator` — a new transform method that converts `tools` from list or comma-separated-string to object format, preserving all other frontmatter and the markdown body.
2. Wires `opencode_agent` format_id into the `integrate_agents_for_target` dispatch so the conversion runs automatically when deploying to `.opencode/agents/`.
3. Adds 9 test cases covering: list→object, comma-separated→object, absent tools, already-object, empty list, missing frontmatter, symlink rejection, body preservation, and end-to-end via target dispatch.

## How to test

```bash
uv run pytest tests/unit/integration/test_agent_integrator.py -v -k OpenCode
```

apm-spec-waiver: Phase 2 of #581 (opencode tools-format transformer); Phase 1 integration path already registered in the opencode_frontmatter module spec boundary

### CODE REVIEW

File: src/apm_cli/integration/agent_integrator.py [hunk 1]
```diff
@@ -365,6 +369,63 @@ def _write_windsurf_agent_skill(
...
+        """
+        if source.is_symlink():
+            raise ValueError(f"Refusing to read symlink source: {source}")
+        content = source.read_text(encoding="utf-8")
+
+        fm_match = AgentIntegrator._FRONTMATTER_RE.match(content)
+        if fm_match:
+            body = content[fm_match.end() :]
+            try:
+                fm = yaml.safe_load(fm_match.group(1)) or {}
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 7a86f645 — added fm normalization with isinstance check + diagnostic warn.

File: src/apm_cli/integration/agent_integrator.py [hunk 2]
```diff
@@ -365,6 +369,63 @@ def _write_windsurf_agent_skill(
...
+            body = content
+            fm = {}
+
+        if not fm_match and not fm:
+            result, links_resolved = self.resolve_links(content, source, target)
+            target.write_text(result, encoding="utf-8")
+            return links_resolved
+
+        tools = fm.get("tools")
+        if tools is not None and not isinstance(tools, dict):
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Same fix as comment 1 — covered by 7a86f645.

File: src/apm_cli/integration/agent_integrator.py [hunk 3]
```diff
@@ -365,6 +369,63 @@ def _write_windsurf_agent_skill(
...
+            if diagnostics is not None:
+                diagnostics.info(
+                    f"Converted tools field from {type(tools).__name__} "
+                    f"to object in {source.name}",
+                )
+
+        fm_yaml = yaml.safe_dump(
+            fm, default_flow_style=False, allow_unicode=True
+        ).rstrip("\n")
+        result = f"---\n{fm_yaml}\n---\n" + body
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 3c5fb975 — docstring now accurately describes what is preserved (key values) and what is not (comments, ordering). The re-serialisation approach is intentional since the fragile alternative (regex-based in-place editing) is more error-prone for the small number of files this touches.

[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Resolved upstream — same fix as comment 3, applied in 3c5fb975.

File: src/apm_cli/integration/agent_integrator.py [hunk 4]
```diff
@@ -365,6 +369,63 @@ def _write_windsurf_agent_skill(
...
+        result, links_resolved = self.resolve_links(result, source, target)
+        target.write_text(result, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 7a86f645 — added target.parent.mkdir(parents=True, exist_ok=True) before write, matching _write_windsurf_agent_skill pattern.

**Comment by futbolsalas15** [comment]
@microsoft-github-policy-service agree company="JobNimbus"

File: src/apm_cli/integration/agent_integrator.py [hunk 5]
```diff
@@ -365,6 +369,66 @@ def _write_windsurf_agent_skill(
...
+            target.write_text(result, encoding="utf-8")
+            return links_resolved
+
+        tools = fm.get("tools")
+        if tools is not None and not isinstance(tools, dict):
+            if isinstance(tools, list):
+                fm["tools"] = {t: True for t in tools if isinstance(t, str)}
+            elif isinstance(tools, str):
+                fm["tools"] = {t.strip(): True for t in tools.split(",") if t.strip()}
+            if diagnostics is not None:
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in c3b62bc3 — diagnostic now lives inside each conversion branch, firing only when actual conversion occurs.

File: src/apm_cli/integration/agent_integrator.py [hunk 6]
```diff
@@ -365,6 +369,66 @@ def _write_windsurf_agent_skill(
...
+                    diagnostics.warn(
+                        f"Non-dict frontmatter in {source.name}, treating as empty",
+                    )
+                fm = {}
+        else:
+            body = content
+            fm = {}
+
+        if not fm_match and not fm:
+            result, links_resolved = self.resolve_links(content, source, target)
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in c3b62bc3 — added target.parent.mkdir(parents=True, exist_ok=True) before write on the early-return path.

File: src/apm_cli/integration/agent_integrator.py [hunk 7]
```diff
@@ -365,6 +369,66 @@ def _write_windsurf_agent_skill(
...
+            target.write_text(result, encoding="utf-8")
+            return links_resolved
+
+        tools = fm.get("tools")
+        if tools is not None and not isinstance(tools, dict):
+            if isinstance(tools, list):
+                fm["tools"] = {t: True for t in tools if isinstance(t, str)}
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in c3b62bc3 — list entries now go through t.strip() + t.strip() guard, consistent with string-form handling. Added plain-whitespace list entry to the test.

File: tests/unit/integration/test_agent_integrator.py [hunk 8]
```diff
@@ -1088,6 +1088,192 @@ def test_sync_integration_opencode_handles_missing_dir(self):
...
+    def test_preserves_no_tools_field(self):
+        """Absent tools field leaves frontmatter unchanged."""
+        source = self.root / "simple.agent.md"
+        source.write_text("---\nname: simple\ndescription: A simple agent\n---\n\n# Simple\n")
+        target = self.root / "simple.md"
+
+        self.integrator._write_opencode_agent(source, target)
+
+        content = target.read_text()
+        assert "tools" not in content
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply tests/unit/integration/test_agent_integrator.py]: Fixed in c3b62bc3 — changed from \"tools\" to \"tools:\" (YAML key pattern) to avoid matching the word in body text or other frontmatter values.

File: src/apm_cli/integration/agent_integrator.py [hunk 9]
```diff
@@ -365,6 +369,71 @@ def _write_windsurf_agent_skill(
...
+            raise ValueError(f"Refusing to read symlink source: {source}")
+        content = source.read_text(encoding="utf-8")
+
+        fm_match = AgentIntegrator._FRONTMATTER_RE.match(content)
+        if fm_match:
+            body = content[fm_match.end() :]
+            try:
+                fm = yaml.safe_load(fm_match.group(1)) or {}
+            except Exception:
+                fm = {}
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py:564]: Fixed in 87a93e11535565746eeff47b0d702cf236af0afa — added diagnostic warning when YAML parsing raises an exception.

File: src/apm_cli/integration/agent_integrator.py [hunk 10]
```diff
@@ -365,6 +369,71 @@ def _write_windsurf_agent_skill(
...
+                    )
+            elif isinstance(tools, str):
+                fm["tools"] = {t.strip(): True for t in tools.split(",") if t.strip()}
+                if diagnostics is not None:
+                    diagnostics.info(
+                        f"Converted tools field from string to object in {source.name}",
+                    )
+
+        fm_yaml = yaml.safe_dump(fm, default_flow_style=False, allow_unicode=True).rstrip("\n")
+        result = f"---\n{fm_yaml}\n---\n" + body
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 87a93e11535565746eeff47b0d702cf236af0afa — introduced conversion_occurred flag; when no conversion is needed the original content passes through verbatim, avoiding re-serialisation of empty or unmodified frontmatter.

File: src/apm_cli/integration/agent_integrator.py [hunk 11]
```diff
@@ -365,6 +369,71 @@ def _write_windsurf_agent_skill(
...
+                fm = {}
+        else:
+            body = content
+            fm = {}
+
+        if not fm_match and not fm:
+            result, links_resolved = self.resolve_links(content, source, target)
+            target.parent.mkdir(parents=True, exist_ok=True)
+            target.write_text(result, encoding="utf-8")
+            return links_resolved
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 87a93e11535565746eeff47b0d702cf236af0afa — removed the early-return branch entirely. All paths (no frontmatter, no conversion, conversion) now converge on a single resolve_links + write tail, eliminating the duplication.

File: tests/unit/integration/test_agent_integrator.py [hunk 12]
```diff
@@ -1088,6 +1088,194 @@ def test_sync_integration_opencode_handles_missing_dir(self):
...
+        import os
+
+        real = self.root / "real.agent.md"
+        real.write_text("---\ntools:\n  - Foo\n---\n\nBody\n")
+        link = self.root / "link.agent.md"
+        os.symlink(real, link)
+        target = self.root / "out.md"
+
+        import pytest
+
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply tests/unit/integration/test_agent_integrator.py]: Fixed in 87a93e11535565746eeff47b0d702cf236af0afa — wrapped os.symlink() in try/except (OSError, NotImplementedError) with pytest.skip, matching the established pattern used throughout the test suite (e.g. test_base_integrator.py, test_unpack_security.py, test_install_scanning.py).

File: src/apm_cli/integration/agent_integrator.py [hunk 13]
```diff
@@ -365,6 +369,76 @@ def _write_windsurf_agent_skill(
...
+                    diagnostics.warn(
+                        f"Failed to parse YAML frontmatter in {source.name}, preserving original",
+                    )
+                fm = {}
+            if not isinstance(fm, dict):
+                if diagnostics is not None:
+                    diagnostics.warn(
+                        f"Non-dict frontmatter in {source.name}, preserving original",
+                    )
+                fm = {}
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 858074db — chose option (b): non-dict frontmatter is preserved as-is (safer, non-destructive). The warning already said "preserving original". Updated test docstring/comment and added assertions that original list items survive.

File: src/apm_cli/integration/agent_integrator.py [hunk 14]
```diff
@@ -365,6 +369,76 @@ def _write_windsurf_agent_skill(
...
+                if diagnostics is not None:
+                    diagnostics.info(
+                        f"Converted tools field from string to object in {source.name}",
+                    )
+
+        if conversion_occurred:
+            fm_yaml = yaml.safe_dump(fm, default_flow_style=False, allow_unicode=True).rstrip("\n")
+            result = f"---\n{fm_yaml}\n---\n" + body
+        else:
+            result = content
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Same fix as sibling comment — both non-dict handling sites are now aligned.

File: tests/unit/integration/test_agent_integrator.py [hunk 15]
```diff
@@ -1088,6 +1088,197 @@ def test_sync_integration_opencode_handles_missing_dir(self):
...
+            os.symlink(real, link)
+        except (OSError, NotImplementedError):
+            pytest.skip("symlink creation not supported on this platform")
+        target = self.root / "out.md"
+
+        with pytest.raises(ValueError, match="symlink"):
+            self.integrator._write_opencode_agent(link, target)
+
+    def test_handles_non_dict_frontmatter(self):
+        """Non-dict YAML frontmatter (e.g. bare list) is treated as empty."""
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply tests/unit/integration/test_agent_integrator.py]: Fixed in 858074db — updated docstring from "treated as empty" to "preserved as-is" and added assertions that - one / - two survive.

File: tests/unit/integration/test_agent_integrator.py [hunk 16]
```diff
@@ -1088,6 +1088,197 @@ def test_sync_integration_opencode_handles_missing_dir(self):
...
+        source = self.root / "badfm.agent.md"
+        # frontmatter that parses as a YAML list, not a mapping
+        source.write_text("---\n- one\n- two\n---\n\n# Body\n")
+        target = self.root / "badfm.md"
+
+        self.integrator._write_opencode_agent(source, target)
+
+        content = target.read_text()
+        # Should not crash; frontmatter becomes empty
+        assert "---" in content
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply tests/unit/integration/test_agent_integrator.py]: Same fix as sibling — assertion strengthened to verify original list items remain.

File: src/apm_cli/integration/agent_integrator.py [hunk 17]
```diff
@@ -365,6 +369,76 @@ def _write_windsurf_agent_skill(
...
+        if source.is_symlink():
+            raise ValueError(f"Refusing to read symlink source: {source}")
+        content = source.read_text(encoding="utf-8")
+
+        fm_match = AgentIntegrator._FRONTMATTER_RE.match(content)
+        if fm_match:
+            body = content[fm_match.end() :]
+            try:
+                fm = yaml.safe_load(fm_match.group(1)) or {}
+            except Exception:
```
[the remark this answers is not in the document]
Comment (futbolsalas15) [reply src/apm_cli/integration/agent_integrator.py]: Fixed in 858074db — narrowed to yaml.YAMLError so unexpected runtime errors still surface.

**Comment by sergio-sisternes-epam** [comment]
## APM Review Panel: `ship_with_followups`

> OpenCode tools-format fix delivers on multi-harness promise; ship with CHANGELOG entry and one missing test as fast follow-ups.

cc @futbolsalas15 @danielmeppiel @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

This PR closes a silent-failure gap that directly contradicts APM's "write once, deploy everywhere" contract. The implementation follows the established Windsurf/Codex transformer pattern cleanly, and all panelists converge on the same signal: no blocking issues, solid unit coverage (15 tests pass), and a small set of polish opportunities.

The most strategically important finding is the devx-ux-expert observation that the failed-YAML-parse path still deploys a broken tools field -- partially reproducing the exact bug this PR fixes. However, this is a pre-existing edge case (malformed YAML in authored agent files is rare), the PR strictly improves the status quo for all well-formed files, and the warn diagnostic does fire. I weight this as a recommended follow-up: the fix should ship now and the edge case hardened in a fast follow-up PR that also adds the missing test the test-coverage-expert flagged.

The python-architect extraction suggestion (`_parse_frontmatter` helper) is sound architecture but premature for this PR -- three call sites is the threshold, and the third just landed here. File it for the next transformer addition.

**Dissent.** No substantive dissent. The devx-ux-expert rated the failed-parse deploy path as "recommended" while test-coverage-expert flagged the same gap as missing test evidence. Both agree on the gap; they differ only on urgency. Ship now, cover it immediately after -- the PR is strictly net-positive even with that edge case unguarded.

**Aligned with:** Multi-harness/multi-host: APM now correctly normalizes tools format for OpenCode, joining Windsurf and Codex as targets with automatic frontmatter adaptation. Portable by manifest: Authors write one agent primitive; APM handles target-specific format requirements at deploy time without modifying the source. OSS community-driven: CHANGELOG entry is missing -- adding it ensures the community sees the fix and trusts the multi-harness surface is actively maintained.

**Growth signal.** OpenCode tools-object fix is a micro-story for the multi-harness narrative: silent deploy failures on target X fixed transparently builds compound trust across the supported-targets table. Ensure the CHANGELOG entry names OpenCode explicitly so it surfaces in release notes.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 1 | 1 | Clean addition following _write_windsurf_agent_skill pattern. No architectural issues. One recommended extraction opportunity for the shared frontmatter-parse boilerplate. |
| CLI Logging Expert | 0 | 0 | 2 | Diagnostic messages are clear, correctly levelled, and consistent with the existing Windsurf pattern. Minor polish opportunities only. |
| DevX UX Expert | 0 | 1 | 1 | Silent auto-conversion is the correct UX choice. One recommended concern: failed YAML parse path still deploys a potentially broken file without a clear user signal. |
| Supply Chain Security Expert | 0 | 0 | 1 | No blocking security issues. Path containment enforced by caller, symlink source rejected, yaml.safe_load/safe_dump prevent injection. |
| OSS Growth Hacker | 0 | 1 | 1 | Fix reinforces the "write once, deploy everywhere" promise. Missing CHANGELOG entry means users will not know it is fixed. |
| Doc Writer | 0 | 2 | 0 | Two doc gaps: CHANGELOG [Unreleased] missing entry; docs/reference/targets-matrix.md omits the tools-format normalization behavior. |
| Test Coverage Expert | 0 | 1 | 0 | Unit coverage solid -- 15 tests pass. One gap: no test for the malformed YAML frontmatter (YAMLError) path. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 5 follow-ups

1. **[OSS Growth Hacker + Doc Writer]** Add CHANGELOG [Unreleased] Fixed entry for OpenCode tools conversion -- User-visible fix with no changelog entry means zero discoverability in release notes. Both panelists flagged independently.
2. **[Test Coverage Expert]** Add `test_malformed_yaml_frontmatter_copies_verbatim` unit test -- Missing test evidence on the YAMLError catch path, the one edge case that partially reproduces the original bug. Ranked above opinion findings per evidence-weighting rules.
3. **[DevX UX Expert]** Emit explicit diagnostic when deploying with potentially invalid tools format after parse failure -- The warn at line 399 is generic; a second message naming the tools-format risk would close the silent-failure loop for the edge case.
4. **[Doc Writer]** Document tools-format normalization in targets-matrix.md -- Users inspecting deployed files will see `{ToolName: true}` instead of their authored list with no explanation. One sentence suffices.
5. **[Python Architect]** Extract `_parse_frontmatter` helper when adding the next target transformer -- Three call sites now share the pattern; extraction belongs in its own PR to keep this diff focused.

### Architecture

```mermaid
classDiagram
    direction LR
    class BaseIntegrator {
        <<Abstract>>
        +resolve_links(content, source, target) tuple
        +copy_agent(source, target) int
    }
    class AgentIntegrator {
        <<TemplateMethod>>
        +integrate_agents_for_target(target, pkg, root) IntegrationResult
        -_write_codex_agent(source, target) None
        -_write_windsurf_agent_skill(source, target, diagnostics) int
        -_write_opencode_agent(source, target, diagnostics) int
        -_FRONTMATTER_RE Pattern
    }
    class TargetProfile {
        <<ValueObject>>
        +format_id str
        +primitives dict
    }
    class IntegrationResult {
        <<ValueObject>>
        +files_integrated int
        +links_resolved int
    }
    class DiagnosticCollector {
        +info(msg)
        +warn(msg)
    }
    BaseIntegrator <|-- AgentIntegrator
    AgentIntegrator ..> TargetProfile : reads format_id
    AgentIntegrator ..> IntegrationResult : returns
    AgentIntegrator ..> DiagnosticCollector : reports to
    classDef touched fill:#fff3b0,stroke:#d47600
    class AgentIntegrator:::touched
```

```mermaid
flowchart TD
    A[integrate_agents_for_target] --> B{mapping.format_id?}
    B -->|codex_agent| C["_write_codex_agent\nread source, write .toml"]
    B -->|windsurf_agent_skill| D["_write_windsurf_agent_skill\nread, resolve_links, write"]
    B -->|opencode_agent| E["_write_opencode_agent\nread source"]
    B -->|default| F["copy_agent\ncopy file"]
    E --> G{tools field type?}
    G -->|list| H["dict comprehension\ntools to object"]
    G -->|str| I["split comma\ndict comprehension"]
    G -->|dict/None| J[preserve original content]
    H --> K["yaml.safe_dump frontmatter"]
    I --> K
    J --> L["resolve_links"]
    K --> L
    L --> M["mkdir + write_text"]
```

### Recommendation

Ship now. The PR eliminates a real silent-failure bug that breaks OpenCode agent deployment for every user who authors tools as a list. All panelists agree: no blocking issues, solid test coverage on the happy path, and the remaining gaps (CHANGELOG, one edge-case test, one doc line) are small enough to land as a same-day follow-up without gating this fix.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[recommended]** Extract shared frontmatter parse + write-back boilerplate into a helper at `src/apm_cli/integration/agent_integrator.py:388`
  Three call sites share symlink guard, read, regex match, yaml.safe_load pattern meeting the Extract Method threshold. A `_parse_frontmatter(source, diagnostics)` helper would deduplicate ~15 lines per transformer and make future targets trivial.
  *Suggested:* Extract `_parse_frontmatter(self, source, diagnostics) -> tuple[dict, str, str]` returning `(frontmatter_dict, body, raw_content)`.

- **[nit]** Type annotation missing on diagnostics parameter at `src/apm_cli/integration/agent_integrator.py:376`
  Semi-public APIs should have type hints. `DiagnosticCollector | None = None` would improve IDE support.

#### CLI Logging Expert

- **[nit]** Warn messages lack `detail=` parameter at `src/apm_cli/integration/agent_integrator.py:399`
  Existing Windsurf warn at line 346 includes a `detail=` string explaining why and what is preserved. New warns omit it. Adding `detail="File will be copied unchanged; fix the frontmatter to enable tools conversion."` matches the existing pattern.

- **[nit]** Verify info messages are suppressed at non-verbose verbosity at `src/apm_cli/integration/agent_integrator.py:420`
  Per-file "Converted tools field..." messages could be noisy at default verbosity when deploying many agents. Confirm DiagnosticCollector suppresses info-level output in non-verbose mode.

#### DevX UX Expert

- **[recommended]** Failed YAML parse path still deploys broken tools format silently at `src/apm_cli/integration/agent_integrator.py:401`
  When YAML parsing fails (lines 396-401), original content is deployed verbatim -- including the broken list-format tools field. This reproduces the exact silent-failure bug the PR aims to fix. The warn diagnostic fires but the deployed file may still be rejected by OpenCode.
  *Suggested:* Add a second diagnostic: "OpenCode: deploying agent with potentially invalid tools format because frontmatter parse failed."

- **[nit]** Key-reordering on conversion not surfaced to the user at `src/apm_cli/integration/agent_integrator.py:432`
  `yaml.safe_dump` alphabetizes keys; the docstring notes this but the user never sees it when inspecting the deployed file.

#### Supply Chain Security Expert

- **[nit]** Target path symlink check not explicit (defense-in-depth) at `src/apm_cli/integration/agent_integrator.py:438`
  Source symlink is checked (line 388) but target is not explicitly verified before write. Upstream `ensure_path_within` covers traversal, but adding a target symlink check would match the defense-in-depth style used for source.

#### OSS Growth Hacker

- **[recommended]** Missing CHANGELOG entry for a user-visible fix at `CHANGELOG.md`
  The [Unreleased] Fixed section has entries but none for this fix. Users scanning the changelog need to see multi-harness fixes to build trust in the deploy-everywhere claim.
  *Suggested:* `- apm install now correctly converts tools list to object when deploying agents to OpenCode, fixing broken .opencode/agents/ files (#1478)`

- **[nit]** Module docstring could briefly mention tools-format normalization at `src/apm_cli/integration/agent_integrator.py:1`
  Brief mention helps future contributors find the conversion logic quickly.

#### Auth Expert -- inactive

No auth-related files changed; PR modifies only the agent file transformer for OpenCode target, which is a local file read/write operation with no auth surface.

#### Doc Writer

- **[recommended]** Missing CHANGELOG entry for tools-format fix at `CHANGELOG.md`
  The [Unreleased] Fixed section has three entries but none for this fix. Suggested: "OpenCode adapter now converts tools lists or comma-separated strings to the `{ToolName: true}` object format required by OpenCode."

- **[recommended]** Deployed tools format normalization undocumented in targets matrix at `docs/src/content/docs/reference/targets-matrix.md`
  The opencode section describes file conventions but omits the tools field transformation. A user who writes `tools: [Read, Glob]` and inspects the deployed file will find `{Read: true, Glob: true}` and may conclude APM mutated their file incorrectly.
  *Suggested:* Add caveat (parallel to existing Cursor and hooks caveats): "Agent tools fields are normalized from list or string form to an object (`{ToolName: true}`) on deploy; required by the OpenCode agent schema. The source primitive is not modified."

#### Test Coverage Expert

- **[recommended]** No test for malformed YAML frontmatter path in `_write_opencode_agent` at `tests/unit/integration/test_agent_integrator.py`
  `yaml.YAMLError` catch at line 397 emits a warn and falls through to verbatim copy. `test_handles_non_dict_frontmatter` covers non-dict parsed YAML but not unparseable YAML. `grep -n "YAMLError|malformed|invalid.*yaml|bad.*yaml"` returned zero hits in the test file.
  *Suggested:* Add `test_malformed_yaml_frontmatter_copies_verbatim`: write `---\n: {bad yaml [\n---\n\n# Body\n` to source, call `_write_opencode_agent`, assert target contains `# Body` and `diagnostics.warn` was called.
  *Proof (missing at unit tier):* `tests/unit/integration/test_agent_integrator.py::test_malformed_yaml_frontmatter_copies_verbatim` -- proves: Malformed YAML in agent frontmatter does not crash deployment and emits a diagnostic warning [multi-harness-support]

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the
`panel-review` label after addressing feedback to re-run.</sub>

**Comment by sergio-sisternes-epam** [comment]
## PR triage recommendation

needs-design

**Recommendation only, not merge or scope approval.** A responsible
human maintainer must approve. Labels, automated advice, and
silence are not approval.

## Linked issue

none -- GitHub closing references are empty. Related: #581 (closed
completed after Phase 1 #1513) and #2483 (closed; maintainer declined
a translator). An issue should have stayed open for Phase 2 after
#581 closed.

## Proposed classification

theme/portability, area/multi-target, type/bug

## Suggested next action

CODEOWNERS @danielmeppiel and @sergio-sisternes-epam should resolve
the recorded policy on intra-file OpenCode translation (#2483, also
#581) before this PR is treated as in-scope. Do not merge from this
comment. Existing review requests already match CODEOWNERS; this
worker will not add or remove them. The head currently conflicts
with main; rebase only after that human scope call.

## Suggested PR comment

```markdown
Thanks @futbolsalas15 -- this is a focused tools list/string-to-object
transform for OpenCode agents, and the inline Copilot threads on this
PR look resolved.

This comment is triage advice only. It is not a diff review and not
merge or scope approval.

Scope is the blocker. #581 was closed when Phase 1 warn-and-copy
shipped in #1513. Maintainer guidance on #581 was that intra-file
translation is too much at this stage, with room for quick wins. On
2026-08-06, #2483 was closed with an explicit record that APM will
not become a vendor translator, and the fuller translation PR #2488
was closed unmerged.

This PR is Phase 2 of #581 without a live accepted issue. It needs a
human call on whether tools-only conversion is a quick-win exception
or whether the #2483 decision applies. The 2026-08-18 review panel
(`ship_with_followups`) is advisory and is not that call. Review
requests already match CODEOWNERS.
```
